proposals/eni.md

```diff
@@ -0,0 +1,129 @@
+<!--
+Copyright 2014-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"). You may
+not use this file except in compliance with the License. A copy of the
+License is located at
+
+     http://aws.amazon.com/apache2.0/
+
+or in the "license" file accompanying this file. This file is distributed
+on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+express or implied. See the License for the specific language governing
+permissions and limitations under the License.
+-->
+### Introduction
+ECS relies on the networking capability provided by Docker to set up the
+network stack for containers. ECS supports the `none` (network interface is not
+configured for any container in the task), `bridge` (a primary network interface
+is configured for containers using by creating a pair of `veth` devices and
+attaching one to the `docker0` bridge and the other to the container) and `host`
+(containers just use host's global default namespace) networking modes. As per
+comments in [this github issue](https://github.com/aws/amazon-ecs-agent/issues/185),
+the `bridge` networking mode results in a number of issues for ECS users:
+
+1. All of the containers launched with this mode share the `docker0` bridge,
+  which results in a **performance penalty**
+2. Since `docker0` bridge is tied to the primary network interface of the
+  instance, it also results in having to configure security groups on
+  the primary network interface to satisfy the requirements of both their tasks
+  and instances (This is also true for containers launched with `host`
+  networking mode), which results in **absence of any network isolation**
+3. The docker `bridge` driver allocates IP Addresses from a private address
+  space to containers and uses NAT rules on the bridge to route packets to
+  containers. Hence, **containers across hosts are not natively addressable**
+  using the IP address allocated to them by docker
+
+Providing EC2-esque network capabilities for containers, where they get their
+own Network Interface that's routable within the VPC, can alleviate many of
+these pain points. Elastic Network Interfaces (ENIs) can be used for this
+purpose. This document details the modifications on the container instance to
+enable the same.
+
+### Overview of the solution
+When an ENI is attached to the container instance for the purpose of being used
```

Comment:
Just noting, that there are [limits for a max number of ENIs per host](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-eni.html#AvailableIpPerENI) which set a bound for a max number of Tasks can be run on a single host. The numbers are relatively small which might lead to low utilization of a host.

How are you going to address the limitations?